daemon/api_sideload_n_try.go

```diff
@@ -260,19 +375,109 @@ func sideloadManySnaps(ctx context.Context, st *state.State, snapFiles []*upload
 		userID = user.ID
 	}
 
-	tss, err := snapstateInstallPathMany(ctx, st, slInfo.sideInfos, slInfo.tmpPaths, userID, &flags.Flags)
+	tss, err := sideloadTaskSets(ctx, st, slInfo, userID, flags.Flags)
 	if err != nil {
-		return nil, errToResponse(err, slInfo.names, InternalError, "cannot install snap files: %v")
+		return nil, err
 	}
 
-	msg := fmt.Sprintf(i18n.G("Install snaps %s from files %s"), strutil.Quoted(slInfo.names), strutil.Quoted(slInfo.origPaths))
-	chg := newChange(st, "install-snap", msg, tss, slInfo.names)
-	chg.Set("api-data", map[string][]string{"snap-names": slInfo.names})
+	snapNames := make([]string, 0, len(slInfo.snaps))
+	snapToComps := make(map[string][]string, len(slInfo.components))
+	for _, sn := range slInfo.snaps {
+		snapName := sn.sideInfo.RealName
+		snapNames = append(snapNames, snapName)
+
+		if len(sn.components) == 0 {
+			continue
+		}
+
+		snapToComps[snapName] = make([]string, 0, len(sn.components))
+		for _, c := range sn.components {
+			snapToComps[snapName] = append(snapToComps[snapName], c.sideInfo.Component.ComponentName)
+		}
+	}
+
+	for _, ci := range slInfo.components {
+		snapToComps[ci.sideInfo.Component.SnapName] = append(snapToComps[ci.sideInfo.Component.SnapName], ci.sideInfo.Component.ComponentName)
+	}
+
+	msg := multiPathInstallMessage(slInfo)
+
+	chg := newChange(st, "install-snap", msg, tss, snapNames)
+	apiData := make(map[string]interface{}, 0)
+
+	if len(snapNames) > 0 {
+		apiData["snap-names"] = snapNames
+	}
+
+	if len(snapToComps) > 0 {
+		apiData["components"] = snapToComps
+	}
+	chg.Set("api-data", apiData)
 
 	return chg, nil
 }
 
-func sideloadSnap(_ context.Context, st *state.State, snapFile *uploadedSnap, flags sideloadFlags) (*state.Change, *apiError) {
+func multiPathInstallMessage(sli *sideloadedInfo) string {
+	var b strings.Builder
+	switch len(sli.snaps) {
+	case 0:
+		b.WriteString(i18n.G("Install"))
+	case 1:
+		b.WriteString(i18n.G("Install snap"))
+	default:
+		b.WriteString(i18n.G("Install snaps"))
+	}
+
+	var paths []string
+	for i, sn := range sli.snaps {
+		fmt.Fprintf(&b, " %q", sn.sideInfo.RealName)
+		paths = append(paths, sn.origPath)
+
+		comps := make([]string, 0, len(sn.components))
+		for _, c := range sn.components {
+			comps = append(comps, c.sideInfo.Component.ComponentName)
+			paths = append(paths, c.origPath)
+		}
+
+		if len(comps) > 0 {
+			b.WriteString(" (")
+			if len(sn.components) > 1 {
+				fmt.Fprintf(&b, i18n.G("with components %s"), strutil.Quoted(comps))
+			} else {
+				fmt.Fprintf(&b, i18n.G("with component %s"), strutil.Quoted(comps))
+			}
+			b.WriteRune(')')
+		}
+
+		if i < len(sli.snaps)-1 {
+			b.WriteRune(',')
+		}
+	}
+
+	compNames := make([]string, 0, len(sli.components))
+	for _, c := range sli.components {
+		compNames = append(compNames, c.sideInfo.Component.String())
+		paths = append(paths, c.origPath)
+	}
+
+	if len(sli.snaps) != 0 && len(sli.components) != 0 {
+		b.WriteString(i18n.G(" and"))
```

Comment:
Side comment, I wonder how well this will work when translating, as the strings are split in many pieces that not all languages will possibly put together in the same way.

Reply:
This was changed in this PR's prereq: https://github.com/canonical/snapd/pull/14623